onnx/defs/traditionalml/defs.cc

```diff
@@ -1098,5 +1120,107 @@ ONNX_ML_OPERATOR_SET_SCHEMA(
+)DOC";
+
+ONNX_ML_OPERATOR_SET_SCHEMA(
+    Tokenizer,
+    2,
+    OpSchema()
+        .SetDoc(Tokenizer_ver2_doc)
+        .Input(0, "X", "UTF-8 strings to tokenize", "tensor(string)")
+        .Output(0, "Y", "UTF-8 Tokenized strings", "tensor(string)")
+        .Attr(
+            "mark",
+            "Boolean whether to mark the beginning/end character with start of text character (0x02)/end of text character (0x03).",
+            AttributeProto::INT)
+        .Attr(
+            "pad_value",
+            "The string used to pad output tensors when the tokens extracted doesn't match the maximum number of tokens found. If start/end markers are needed, padding will appear outside the markers.",
+            AttributeProto::STRING)
+        .Attr(
+            "tokenexp",
+            "An optional string. Token's regular expression in ECMA format (http://ecma-international.org/ecma-262/5.1/#sec-15.10)."
+            " If set, tokenizer may produce tokens matching the specified pattern. Note that one and only of \"tokenexp\" and"
+            "\"separators\" should be set.",
+            AttributeProto::STRING,
+            OPTIONAL)
+        .Attr(
+            "separators",
+            "an optional list of strings attribute that contains a list of separators - regular expressions to match separators"
+            " Two consecutive segments in X connected by a separator would be divided into two tokens."
+            " For example, if the input is \"Hello World!\" and this attribute contains only one space character,"
+            " the corresponding output would be [\"Hello\", \"World!\"]. To achieve character-level tokenization,"
+            " one should set the 'separators' to [\"\"], which contains an empty string.",
+            AttributeProto::STRINGS,
+            OPTIONAL)
+        .Attr(
+            "mincharnum",
+            "Minimum number of characters allowed in the output. For example, if mincharnum is 2, tokens such as \"A\" and \"B\" would be ignored",
+            AttributeProto::INT)
+        .TypeAndShapeInferenceFunction([](InferenceContext& ctx) {
+          propagateElemTypeFromInputToOutput(ctx, 0, 0);
+          if (!hasInputShape(ctx, 0)) {
+            return;
+          }
+          TensorShapeProto output_shape;
+          auto& input_shape = ctx.getInputType(0)->tensor_type().shape();
+          const auto dim_size = input_shape.dim_size();
+          if (dim_size == 1) {
+            // Input is [C], Output is [C][?]
+            // Copy C
+            *output_shape.add_dim() = input_shape.dim(0);
+            // Add unknown D
+            output_shape.add_dim();
+          } else if (dim_size == 2) {
+            // Input is [N, C], Output is [N][C][?]
+            // Copy N, C
+            *output_shape.add_dim() = input_shape.dim(0);
+            *output_shape.add_dim() = input_shape.dim(1);
+            // Add unknown D
+            output_shape.add_dim();
+          }
```

Comment:
Copy (local) output_shape to output, or make it a reference. (But the reference would be a problem if neither branch is taken, since it will appear as though output-shape has rank 0.)

Reply:
If rank > 2, may be it would be better (for debugging) to throw a shape-inference exception.